signer/threshold_validator.go

```diff
@@ -347,27 +348,26 @@ func (pv *ThresholdValidator) waitForPeerNonces(
 	peer Cosigner,
 	hrst HRSTKey,
 	wg *sync.WaitGroup,
-	encryptedNoncesThresholdMap map[Cosigner][]CosignerNonce,
+	nonces map[Cosigner][]CosignerNonce,
 	thresholdPeersMutex *sync.Mutex,
 ) {
 	peerStartTime := time.Now()
-	ephemeralSecretParts, err := peer.GetNonces(chainID, hrst)
+	peerNonces, err := peer.GetNonces(chainID, hrst)
 	if err != nil {
-
 		// Significant missing shares may lead to signature failure
 		missedNonces.WithLabelValues(peer.GetAddress()).Add(float64(1))
 		totalMissedNonces.WithLabelValues(peer.GetAddress()).Inc()
-		pv.logger.Error("Error getting secret parts", "peer", peer.GetID(), "err", err)
+		pv.logger.Error("Error getting nonces", "cosigner", peer.GetID(), "err", err)
 		return
 	}
 	// Significant missing shares may lead to signature failure
 	missedNonces.WithLabelValues(peer.GetAddress()).Set(0)
 	timedCosignerNonceLag.WithLabelValues(peer.GetAddress()).Observe(time.Since(peerStartTime).Seconds())
 
-	// Check so that getEphemeralWaitGroup.Done is not called more than (threshold - 1) times which causes hardlock
+	// Check so that wg.Done is not called more than (threshold - 1) times which causes hardlock
 	thresholdPeersMutex.Lock()
-	if len(encryptedNoncesThresholdMap) < pv.threshold-1 {
-		(encryptedNoncesThresholdMap)[peer] = ephemeralSecretParts.Nonces
+	if len(nonces) < pv.threshold-1 {
+		nonces[peer] = peerNonces.Nonces
 		defer wg.Done()
 	}
```

Comment:
This is old code but `defer wg.Done()` conditionally (after an `if`) makes me thing there's a chance for a deadlock. Hopefully something else triggers `wg.Done()`.

Reply:
this is done intentionally, It does `wg.Add(pv.threshold-1)` because we want to stop waiting once we have threshold responses (local cosigner included for -1). The defer is used to make sure the wg.Done happens after unlocking the mutex.

Reply:
Alternate approach using channels instead of a mutex:

```go
var wg sync.WaitGroup
wgCh := make(chan *sync.WaitGroup, threshold)
for i := 0; i < threshold; i++ {
  // Seed only the threshold count worth of wait groups,
  // so we can wait on the threshold being reached, not the total.
  wgCh <- &wg
}

wg.Add(threshold)
for i := 0; i < total; i++ {
  go foo(wgCh)
}

// func foo(wgCh chan *sync.WaitGroup):
defer func() {
  if wg, ok := <-wgCh; ok {
    // Haven't called wg.Done on the threshold count yet.
    wg.Done()
  }
}
```